src/ert/services/ert_server.py

```diff
@@ -0,0 +1,310 @@
+            )
+            raise pe
+        return ErtServerContext(service)
+
+    def fetch_url(self) -> str:
+        """Returns the url. Blocks while the server is starting"""
+        if self._url is not None:
+            return self._url
+
+        for url in self.fetch_connection_info()["urls"]:
+            con_info = self.fetch_connection_info()
+            try:
+                resp = requests.get(
+                    f"{url}/healthcheck",
+                    auth=self.fetch_auth(),
+                    verify=con_info["cert"],
+                )
+                logging.getLogger(__name__).info(
+                    f"Connecting to {url} got status: "
+                    f"{resp.status_code}, {resp.headers}, {resp.reason}, {resp.text}"
+                )
+                if resp.status_code == 200:
+                    self._url = url
+                    return str(url)
+
+            except requests.ConnectionError as ce:
+                logging.getLogger(__name__).info(
+                    f"Could not connect to {url}, but will try something else. "
+                    f"Error: {ce}"
+                )
+        raise TimeoutError(
+            "None of the URLs provided for the ert storage server worked."
+        )
+
+    @classmethod
+    def session(cls, project: os.PathLike[str], timeout: int | None = None) -> Client:
+        """
+        Start a HTTP transaction with the server
+        """
+        inst = cls.connect(timeout=timeout, project=project)
+        info = inst.fetch_connection_info()
+        return Client(
+            conn_info=ErtClientConnectionInfo(
+                base_url=inst.fetch_url(),
+                auth_token=inst.fetch_auth()[1],
+                cert=info["cert"],
+            )
+        )
+
+    @property
+    def logger(self) -> logging.Logger:
+        return logging.getLogger("ert.shared.storage")
+
+    def shutdown(self) -> int:
+        """Shutdown the server."""
+        if self._thread_that_starts_server_process is None:
+            return -1
+
+        self.__class__._instance = None
+        error_code = self._thread_that_starts_server_process.shutdown()
+        self._thread_that_starts_server_process = None
+
+        return error_code
+
+    @classmethod
+    def connect(
+        cls,
+        *,
+        project: os.PathLike[str],
+        timeout: int | None = None,
+        logging_config: str | None = None,
+    ) -> ErtServer:
+        if cls._instance is not None:
+            cls._instance.wait_until_ready()
+            assert isinstance(cls._instance, cls)
+            return cls._instance
+
+        path = Path(project)
+
+        # Wait for storage_server.json file to appear
+        try:
+            if timeout is None:
+                timeout = 240
+            t = -1
+            while t < timeout:
+                if (path / "storage_server.json").exists():
+                    with (path / "storage_server.json").open() as f:
```

Comment:
this pattern is probably susceptible to a race condition if the storage_server.json file is touched before written to. Should probably both check it exists and has non-zero content before returning ErtServer.

Unless we have some special meaning behind keeping the file descriptor `f` open during the construction of `ErtServer`, we should read the file first (using pathlib) and then make the ErtServer object.